keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+requests made with that class:
+
+```yaml
+kind: ResourceClassParameters
+apiVersion: resource.k8s.io/v1alpha2
+
+metadata:
+  name: someArbitraryName
+
+generatedFrom:
+  name: gpu-parameters
+  kind: CardClassParameters
+  apiGroup: dra.example.com
+  uid: foobar-uid
+
+vendorParameters:
+  ...
+
+filters:
+- driverName: cards.dra.example.com
+  partitionable:
+    selector: |-
+      quantities["memory"] <= "16Gi"
+```
+
+In this example, the additional selector expression limits users of this class
+to smaller sub-partitions. Together with limiting the number of claims that
+users are allowed to create for this class (see resource quotas in the core
+KEP) this can ensure that users do not consume too much resources. Allowing
+resource quotas that are based on resource attributes may be a useful future
+enhancement.
+
+### Communicating allocation to the DRA driver
+
+The scheduler decides which resources to use for a claim and how much of
+them. It also needs to pass through the opaque vendor parameters, if there are
+any. All of this information gets stored in the allocation result inside the
+ResourceClaim status. For the example above, the result produced by the
+scheduler is simply the list of IDs of the selected sub-partitions:
+
+```yaml
+# Matches with the SemanticResourceHandle Go type defined below.
+vendorClassParameters:
+  ...
+vendorClaimParameters:
+  kind: CardParameters
+  apiVersion: dra.example.com/v1alpha1
+  metadata:
+    name: my-parameters
+    namespace: user-namespace
+    uid: foobar-uid
+  ...
+  spec:
+    count: 2
+    minimumRuntimeVersion: v12.0.0
+    minimumMemory: 8Gi
+    sharing:
+      strategy: TimeSliced
+
+nodeName: worker-1
+partitionable:
+  instances:
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-0
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-1
+```
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```go
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via semantic parameters, then SemanticParameters must be set to true.
+    SemanticParameters bool
+}
+```
+
+### NodeResourceSlice
+
+For each node, one or more NodeResourceSlice objects get created. The kubelet
+publishes them with the node as the owner, so they get deleted when a node goes
+down and then gets removed.
+
+All list types are atomic because that makes tracking the owner for
+server-side-apply (SSA) simpler. Patching individual list elements is not
+needed and there is a single owner (kubelet).
+
+```go
+// NodeResourceSlice provides information about available
+// resources on individual nodes.
+type NodeResourceSlice struct {
+    metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    Spec NodeResourceSliceSpec
```

Comment:
API reviewer question: is "spec" really the right thing here? This is not user intent, but rather simply discovered data about the node.  Should this just be `data` rather than `spec`?

Reply:
Not that is influenced by this spec. This is really "what is the raw capacity if there are no allocations". The status will be "what is the current capacity / current allocations" or something to that effect. The "allocation intent" is captured in the allocation results that are stored in individual resource claim statuses.

Reply:
Where do we store "gpo-0 is currently configured in mode X; gpu-1 is currently configured in mode Y" ?

Reply:
I don't think in the simplified model we are doing modes, so I think we can sort of gloss over this now.

But the answer right now would be in the the allocation results. Those are associated with claims. So, each claim that uses a card in a mode (not just the first), needs to record that. Then, once all those claims are gone, we are back down to "zero" and the card is reconfigurable again.

This may not be ideal. I suspect we might want a more complete model. But I don't think we need to solve that now.